core/src/main/java/com/linecorp/armeria/common/util/CountingSampler.java

```diff
@@ -46,23 +48,22 @@
  *
  * <p>Forked from brave-core 5.6.3 at d4cbd86e1df75687339da6ec2964d42ab3a8cf14
  */
-final class CountingSampler extends Sampler {
+final class CountingSampler implements Sampler {
 
     /**
      * Creates a new instance.
      *
-     * @param rate 0 means never sample, 1 means always sample. Otherwise minimum sample rate is 0.01,
-     *             or 1% of traces
+     * @param rate {@code 0.0} means never sample, {@code 1.0} means always sample. Otherwise minimum sampling
+     *             rate is between {@code 0.01} and {@code 1.0}.
      */
-    public static Sampler create(final float rate) {
-        if (rate == 0) {
-            return NEVER_SAMPLE;
-        }
-        if (rate == 1.0) {
-            return ALWAYS_SAMPLE;
+    static Sampler create(final double rate) {
+        final int percent = (int) (rate * 100.0);
+        checkArgument(percent >= 0 && percent <= 100, "rate should be between 0.0 and 1: was " + rate);
+        if (percent == 0) {
+            return Sampler.never();
         }
-        if (rate < 0.01f || rate > 1) {
-            throw new IllegalArgumentException("rate should be between 0.01 and 1: was " + rate);
+        if (percent == 100) {
+            return Sampler.always();
         }
         return new CountingSampler(rate);
```

Comment:
I think we can change the constructor to take an integer value which is `percent` because the rate is changed to percent after all. :)

Reply:
so less than a percent isn't useful? mind you that in zipkin less than a percent is common

Reply:
nevermind me :) I recall we used different sampler when the percentage was below 1%

Reply:
Yes, the previous logic was ignoring that anyway. :) Which sampler Zipkin uses when the percentage was below 1%? I think we should follow that logic. 🤔

Reply:
The other sampler won't work for you because it is based on a pattern on random ID (Boundary Sampler). High rate can be addressed by switching to rate-based, or you can go with counting with another layer of bucket granularity. The lowest sample rate supported was .01pct and I think some were doing .03% Not sure the translation for logging, but that's the context in tracing.

Reply:
Then let's keep the sampler as it is and make another one for the rate under 0.01 when there's a demand for that.
Thanks for the explanation. 😄

Reply:
@minwoox Thank you for review. I missed this because I didn't see deep to the constructor.
Fixed in cd47b3b

@adriancole Thank you for the additional information and explanation.